src/cpp/common/tls_credentials_options.cc

```diff
@@ -0,0 +1,213 @@
+    const grpc::string& target_name) {
+  c_arg_->target_name = gpr_strdup(target_name.c_str());
+}
+
+void TlsServerAuthorizationCheckArg::set_peer_cert(
+    const grpc::string& peer_cert) {
+  c_arg_->peer_cert = gpr_strdup(peer_cert.c_str());
+}
+
+void TlsServerAuthorizationCheckArg::set_status(grpc_status_code status) {
+  c_arg_->status = status;
+}
+
+void TlsServerAuthorizationCheckArg::set_error_details(
+    const grpc::string& error_details) {
+  c_arg_->error_details = gpr_strdup(error_details.c_str());
+}
+
+void TlsServerAuthorizationCheckArg::OnServerAuthorizationCheckDoneCallback() {
+  c_arg_->cb(c_arg_);
+}
+
+/** gRPC TLS server authorization check config API implementation **/
+TlsServerAuthorizationCheckConfig::TlsServerAuthorizationCheckConfig(
+    const void* config_user_data,
+    int (*schedule)(void* config_user_data,
+                    TlsServerAuthorizationCheckArg* arg),
+    void (*cancel)(void* config_user_data, TlsServerAuthorizationCheckArg* arg),
+    void (*destruct)(void* config_user_data))
+    : config_user_data_(const_cast<void*>(config_user_data)),
+      schedule_(schedule),
+      cancel_(cancel),
+      destruct_(destruct) {
+  c_config_ = grpc_tls_server_authorization_check_config_create(
+      config_user_data_, &TlsServerAuthorizationCheckConfigCSchedule,
+      &TlsServerAuthorizationCheckConfigCCancel, destruct_);
+  c_config_->set_context(static_cast<void*>(this));
+}
+
+TlsServerAuthorizationCheckConfig::~TlsServerAuthorizationCheckConfig() {
+  ::grpc_core::Delete(c_config_);
+}
+
+/** gRPC TLS credential options API implementation **/
+TlsCredentialsOptions::TlsCredentialsOptions(
+    grpc_ssl_client_certificate_request_type cert_request_type,
+    std::shared_ptr<TlsKeyMaterialsConfig> key_materials_config,
+    std::shared_ptr<TlsCredentialReloadConfig> credential_reload_config,
+    std::shared_ptr<TlsServerAuthorizationCheckConfig>
+        server_authorization_check_config)
+    : cert_request_type_(cert_request_type),
+      key_materials_config_(std::move(key_materials_config)),
+      credential_reload_config_(std::move(credential_reload_config)),
+      server_authorization_check_config_(
+          std::move(server_authorization_check_config)) {
+  c_credentials_options_ = grpc_tls_credentials_options_create();
+  grpc_tls_credentials_options_set_cert_request_type(c_credentials_options_,
+                                                     cert_request_type_);
+  grpc_tls_credentials_options_set_key_materials_config(
+      c_credentials_options_,
+      ConvertToCKeyMaterialsConfig(key_materials_config_));
+  grpc_tls_credentials_options_set_credential_reload_config(
+      c_credentials_options_, credential_reload_config_->c_config());
+  grpc_tls_credentials_options_set_server_authorization_check_config(
+      c_credentials_options_, server_authorization_check_config_->c_config());
+}
+
+TlsCredentialsOptions::~TlsCredentialsOptions() {
+  gpr_free(c_credentials_options_);
```

Comment:
Could you please use `Delete` here instead of `gpr_free`?

Reply:
Actually I can't use Delete here - doing so creates an error in the `TlsCredentialsOptionsCppToC` unit test. 

The reason is the following: the user creates `TlsKeyMaterialsConfig`, `TlsCredentialReloadConfig`, and/or `TlsServerAuthorizationCheckConfig` objects, and then passes them to the constructor of `TlsCredentialsOptions`. The constructor assigns the underlying C config of `TlsCredentialReloadConfig` is used to instantiate `c_credentials_options_` (and similarly for server authorization). 
Now, if the destructor of `TlsCredentialsOptions` uses `Delete`, then it frees the C config underlying `TlsCredentialReloadConfig`. Then, since the `TlsCredentialReloadConfig` object was created prior the `TlsCredentialsOptions` object , its destructor is still called and it tries to free the C config again, hence the error.

Reply:
Ah I see. That makes sense. Thanks for the explanation, Matt.